src/coreclr/src/vm/diagnosticsprotocol.h

```diff
@@ -103,6 +103,56 @@ namespace DiagnosticsIpc
 
     const MagicVersion DotnetIpcMagic_V1 = { "DOTNET_IPC_V1" };
 
+    /**
+     * ==ADVERTISE PROTOCOL==
+     * Before standard IPC Protocol communication can occur on a client-mode connection
+     * the runtime must advertise itself over the connection.  ALL SUBSEQUENT COMMUNICATION 
+     * IS STANDARD DIAGNOSTICS IPC PROTOCOL COMMUNICATION.
+     * 
+     * See spec in: dotnet/diagnostics@documentation/design-docs/ipc-spec.md
+     * 
+     * The flow for Advertise is a one-way burst of 24 bytes consisting of
+     * 8 bytes  - "ADVR_V1\0" (ASCII chars + null byte)
+     * 16 bytes - random 128 bit number cookie (little-endian)
+     * 8 bytes  - PID (little-endian)
+     */
+
+    const uint8_t AdvertiseMagic_V1[8] = "ADVR_V1";
+
+    const uint32_t AdvertiseSize = 32;
+
+    static GUID AdvertiseCookie_V1 = GUID_NULL;
+
+    inline GUID GetAdvertiseCookie_V1()
+    {
+        if (AdvertiseCookie_V1 == GUID_NULL)
+        {
+            CoCreateGuid(&AdvertiseCookie_V1);
+        }
+
+        return AdvertiseCookie_V1;
+    }
+
+    inline bool SendIpcAdvertise_V1(IpcStream *pStream)
+    {
+        uint8_t advertiseBuffer[DiagnosticsIpc::AdvertiseSize];
+        GUID cookie = GetAdvertiseCookie_V1();
+        uint64_t pid = GetCurrentProcessId();
+
+        uint64_t *buffer = (uint64_t*)advertiseBuffer;
+        buffer[0] = *(uint64_t*)AdvertiseMagic_V1;
+        buffer[1] = (((uint64_t)VAL32(cookie.Data1) << 32) | ((uint64_t)VAL16(cookie.Data2) << 16) | VAL16((uint64_t)cookie.Data3));
+        buffer[2] = *(uint64_t*)cookie.Data4;
+        buffer[3] = VAL64(pid);
+
+        uint32_t nBytesWritten = 0;
+        if (!pStream->Write(advertiseBuffer, sizeof(advertiseBuffer), nBytesWritten, 100 /* ms */))
```

Comment:
Currently Write() treats partial writes as being successful but probably most callers aren't handling that possibility. Rather than making every caller handle the possibility of partial writes I'd recommend putting a loop inside Write() so that the only possible outcomes are complete writes or errors.

Ditto for Read()

Reply:
The Windows implementations use overlapped IO which notifies us only if the write is complete (in success or failure):
> If the bWait parameter is TRUE, GetOverlappedResult determines whether the pending operation has been completed by waiting for the event object to be in the signaled state.

I added a loop inside the unix implementations that recalls the APIs if they were interrupted/didn't complete the total number of bytes.